nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp

```diff
@@ -272,18 +318,44 @@ class BtActionNode : public BT::ActionNodeBase
         }
```

Comment:
Should't we check if the `goal_handle_` is `nullptr`? because it seems that you call `goal_handle_.reset()` in the timeout handling segment.

Reply:
If we get a `result`, there shouldn't be any situation where the handle is a nullpointer, this is a callback on success / failure / cancel. If its a nullpointer, we have an issue where we can't then get the `get_goal_id()` in order to validate that this is the proper response to set the flags that the action has completed to return a result to the behavior tree in `tick()`. 

If this is an issue, we need to rethink how we deal with the goal handles in the bt action node.

The `goal_handle_` I don't think is set to anything when it is reset due to timeouts. The future to the goal handle (https://github.com/ros-planning/navigation2/blob/f83612d28ce0db76436cf711071129eb17099c8f/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L321) is still spinning to get valid. The `goal_handle_` from the future isnt set until it spins and gets back a successful goal acceptance (https://github.com/ros-planning/navigation2/blob/f83612d28ce0db76436cf711071129eb17099c8f/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L351) so I think in (all?) most cases, that should be a nullpointer anyway.

@naiveHobo if that is correct, should the checks on `goal_handle_` in some places actually be `future_goal_handle_` and we reset `future_goal_handle_` on timeout? I think we may have been mixing these up at various points. Should it actually be:

```
    if (future_goal_handle_) {
      auto elapsed = (node_->now() - time_goal_sent_).to_chrono<std::chrono::milliseconds>();
      if (!is_future_goal_handle_complete(elapsed)) {
        // return RUNNING if there is still some time before timeout happens
        if (elapsed < server_timeout_) {
          return BT::NodeStatus::RUNNING;
        }
        // if server has taken more time to respond than the specified timeout value return FAILURE
        RCLCPP_WARN(
          node_->get_logger(),
          "Timed out while waiting for action server to acknowledge goal request for %s",
          action_name_.c_str());
        future_goal_handle_.reset();
        return BT::NodeStatus::FAILURE;
      }
    }
```

?? We track and reset the `future_goal_handle_` instead? Before we set, we should default the `future_goal_handle_` to nullpointer. Then if we send a goal, it becomes non-null to then spin the future. Once the future is complete (or times out or fails) we should reset the `future_goal_handle_` instead back to nullpointer? I'm not sure anything relating to `send_goal` and `is_future_goal_handle_complete` needs `goal_handle_`, I think it needs `future_goal_handle_` checks, yeah?

Reply:
I think if we remove the `goal_handle_.reset()`'s for this logic in `tick()` and `is_future_goal_handle_complete()`, then the only remaining ones are in `halt` and the one at the end of `tick()` once a final solution is available. **Would that make things work well?** The other 4 could be the `future_goal_handle_.reset()` I think...

(this is overall a subtle issue so correct me if I missed something)

Reply:
At the end of `tick()` I agree, because we wouldn't have gotten there without a `result` in that callback. Though at the end of `halt()` are we _guaranteed_ to have received a result? I'm actually not 100% sure about that. Does the `spin_until_future_complete` spin until the action server accepts the cancellation or until the cancellation is actually complete? 

If it just waits for the cancel to be accepted, in the simple action server (https://github.com/ros-planning/navigation2/blob/9f0e85f312770a63eeac5b37fc4b8745c38187d3/nav2_util/include/nav2_util/simple_action_server.hpp#L135-L149) that doesn't promise that we have actually completed that cancel (I think that happens in https://github.com/ros-planning/navigation2/blob/9f0e85f312770a63eeac5b37fc4b8745c38187d3/nav2_util/include/nav2_util/simple_action_server.hpp#L523 `canceled`, probably)

As such, maybe its best for the goal handle reset to be in end of tick and in the result callback after the quantities are set if the goal IDs match?

Reply:
Yup, got it. I'll remove the reset at the end of `halt()`. Although, I don't think we should reset the goal handle in the result callback. It might be useful to have access to the last goal through the goal handle in the `on_success()`, `on_aborted()`, `on_cancelled()` functions after the goal result is available. The reset at the end of `tick()` should ensure that the goal handle is always reset after a goal.